File: src/main/java/me/creepinson/creepinoutils/api/util/threading/Worker.java
package me.creepinson.creepinoutils.api.util.threading;

import java.util.concurrent.atomic.AtomicBoolean;

public abstract class Worker extends Thread {
    private final AtomicBoolean shouldWait = new AtomicBoolean();

    protected abstract boolean processingIsComplete();

    protected abstract void process();

    protected abstract void cleanUpResources();

    public abstract Object getLock();

    public void disable() {
        shouldWait.set(false);
    }

    public void enable() {
        shouldWait.set(true);
    }

    @Override
    public void run() {
        try {
            while (!processingIsComplete()) {
                while (!shouldWait.get()) {
                    synchronized (getLock()) {
                        getLock().wait();
                    }
                }
            }
            process();
        } catch (InterruptedException e) {
            System.out.println("Worker thread stopped");
        } finally {
            cleanUpResources();
        }
    }
}

File: src/main/java/dev/throwouterror/util/worker/Worker.kt
package dev.throwouterror.util.worker

import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

abstract class Worker : Thread() {
    private val shouldWait = AtomicBoolean()
    protected abstract fun processingIsComplete(): Boolean
    protected abstract fun process()
    protected abstract fun cleanUpResources()
    val lock: ReentrantLock = ReentrantLock()

    fun disable() {
        shouldWait.set(false)
    }

    fun enable() {
        shouldWait.set(true)
    }

    override fun run() {
        try {
            val condition = lock.newCondition()
            while (!processingIsComplete()) {
                while (!shouldWait.get()) {
                    lock.withLock {
                        condition.await()
                    }
                }
            }
            process()
        } catch (e: InterruptedException) {
            println("Worker thread stopped")
        } finally {
            cleanUpResources()
        }
    }
}